### DISCUSSION TITLE
Alternative to events for tracing AsyncEngine and AsyncSession

### DISCUSSION THREAD

**Discussion by mnbbrown** [sqlalchemy/sqlalchemy#6717] [Usage Questions] [answered]
Currently OpenTracing [1] and OpenTelemetry [2] use the event system for tracing sqlalchemy. With the event system not being available with async sessions and engines [3] yet can anybody suggest a suitable alternative approach? 

[1] https://github.com/opentracing-contrib/python-sqlalchemy
[2] https://github.com/open-telemetry/opentelemetry-python-contrib/tree/main/instrumentation/opentelemetry-instrumentation-sqlalchemy
[3] https://github.com/sqlalchemy/sqlalchemy/blob/master/lib/sqlalchemy/ext/asyncio/events.py

**Comment by CaselIT** [comment, marked as answer]
Hi,

Using the proxied object, like suggested here https://github.com/sqlalchemy/sqlalchemy/discussions/6594, is not an option?

**Reply by mnbbrown** [reply]
Perfect! I should have found that thread before asking - apologies. Now to figure out how to make the instrumentation libs support both sync and async seamlessly :)

**Reply by CaselIT** [reply]
I think you could just check for the type of engine/connection/session and use the sync version or not. 

It may be useful to have a flag `is_async=True/False` on engine/connection/session/transaction/.... What do you think @zzzeek ?

**Reply by zzzeek** [reply]
im not into that idea right now, it's kind of vague what that means.    stick to isinstance() for now.

**Reply by mnbbrown** [reply]
I went for the more explicit option of passing the `sync_engine` explicitly: https://github.com/open-telemetry/opentelemetry-python-contrib/pull/568/files

**Reply by mnbbrown** [reply]
One small follow up question. Was there a change in 1.4 around how add_all works for the postgres driver? For most drivers it seems to generate individual `INSERT` queries but the postgres driver generates one `INSERT` with multiple tuples.

**Reply by mnbbrown** [reply]
Actually - it doesn't appear to be isolated to 1.4. I will do some more digging.

**Reply by CaselIT** [reply]
Yes, an optimization was enabled by default. See https://docs.sqlalchemy.org/en/14/changelog/migration_14.html#orm-batch-inserts-with-psycopg2-now-batch-statements-with-returning-in-most-cases
